### ISSUE TITLE
called functions not highlighted

### ISSUE THREAD

**Issue by andykais** [pangloss/vim-javascript#410]
after looking through previous issues, like [this one](https://github.com/pangloss/vim-javascript/issues/406), it seems that functions should be highlighted in all cases, and be the same color. Called functions are not for my editor however. Running vim-gtk 7.4 (2013 Aug 10, compiled Jan  2 2014 19:39:59). Colorscheme [seoul256](https://github.com/junegunn/seoul256.vim). I have confirmed the same problem exists when no colorscheme is present.

![screenshot 59](https://cloud.githubusercontent.com/assets/6627147/14939532/9d6165fe-0f4b-11e6-9f80-e6f5642ba47d.png)

**Comment by amadeus** [comment]
So a couple things. What branch are you on? If you are on `develop`, make sure you are up to date.

Function definitions are not highlighted the same as function calls.

Function calls are not actually linked to anything, and would need to manually highlight them using the `jsFuncCall` group.

**Comment by andykais** [comment]
so this is normal behavior? I am using the vim-plug package manager and I can confirm the result is the same on both `master` and `develop` branch. It seems like this should be a normal behaviour, Atom, for example, has this behavior by default.

![screenshot 60](https://cloud.githubusercontent.com/assets/6627147/14941518/16389e68-0f9f-11e6-94ac-ab855bb668ef.png)

If I need to define it myself, can you show me a line I could add to my .vimrc or where I can find a config example that does this?

**Comment by andykais** [comment]
I have a solution to my initial problem, which is to highlight the `jsFuncCall` group as you mentioned. This is not a good solution however, because that will overwrite any other highlighting of groups like `jsBuiltins` I personally define, like node.js builtins:

```
autocmd FileType javascript syntax keyword jsBuiltins watch require
```

 or groups defined from [vim-javascript-syntax](https://github.com/jelera/vim-javascript-syntax) such as `document.getElementById()`


| Without highlighting `jsFuncCall` | With highlighting `jsFuncCall` |
:---------------------------------------------:|:----------------------------------------:
![screenshot 65](https://cloud.githubusercontent.com/assets/6627147/15011677/70357480-11f3-11e6-878d-f3cf76c4dd15.png) | ![screenshot 66](https://cloud.githubusercontent.com/assets/6627147/15011736/b887712a-11f3-11e6-8189-2696136040e2.png)


I see that from this [stack post](http://stackoverflow.com/questions/29192124/how-to-color-function-call-in-vim-syntax-highlighting) function *definitions* are usually highlighted but not function *calls*, still I think this should be built into `vim-javascript`

An alternate solution I see, is to use [easytags](http://peterodding.com/code/vim/easytags/) or [tern](https://github.com/ternjs/tern_for_vim) to create tags for just about everything in a file, but that appears to be doing much more than highlighting, can be memory intensive, and has a steep learning curve.

**Comment by amadeus** [comment]
You're bringing up a valid point, however I am not sure that it is within scope of this project. Ultimately the way Vim syntax highlighting works is through very basic regexes. Its goal is not to necessarily parse out crazy context aware patterns in your JS code, like you provided examples of above, but to simply match based on commonly used patterns. There are many limits and exceptions to this, and it won't always be 100% correct, but at least we are close and 'usually' correct.

For example, we could look into changing the priority of `jsFuncCall`, to be lower in the context chain relative to `require`, but then we ultimately still get the issue if you use the `require` keyword in another context, say within a JS object, e.g. `obj.require`. Technically it's no longer the global `require` method, and now we've introduced a slippery slope. We could make the regex for matching require stronger, to NOT match that context, but that requires look-behind pattern matching that will kill Vim performance.

There's a valid argument around 'correctness' as well. Technically, `require` is being used as a function call here, and the intention of `jsFuncCall` is to simply tell you, the developer, that you are are executing a function. It knows nothing about whether you previously defined said function or not.

Ultimately however, I would love it if Vim would allow us to asynchronously use a JS parser to actually create syntax matching groups, however until then, it's going to be fairly simplistic, and probably won't exactly match certain things you may be used to in Sublime or Atom. At the end of the day though, this is simply differences in intended direction by the plugin maintainers, and if we make a change for this, someone else may start an issue about a potentially valid and fair alternate direction.

In summary, I will close this issue for now, however I am open to re-opening it up if there is more dialogue that could convince myself or other maintainers otherwise.

**Comment by andykais** [comment]
I understand. I assume that async calls are how Atom and Sublime achieve this superior syntax highlighting. In the interest of future plans, [Vim 8.0](https://github.com/vim/vim/blob/master/runtime/doc/version8.txt) is coming, which supposedly supports asynchronous operations natively. Do you have any more complex syntax support plans once it arrives? 

Its a definite advantage Atom has over vim, especially supporting this syntax out of the box, as opposed to the complex setup and customization I need if I want to create this behavior in my vim environment.

**Comment by amadeus** [comment]
Not sure how Atom works, but no, I don't believe Atom and Sublime do these things asynchronously. 

I wouldn't say they are superior either. We have for more granularity in many places. As I said earlier, it's mostly differences in opinion. AFAIK, only Emacs can do the crazy parser stuff (perhaps some IDEs as well?).

Vim introducing async stuff might allow for some workaround hacks, but I don't think it could reasonably be used for syntax highlighting. They would need to build it into the syntax API.

And to your final point, Vim is an interface that was designed to be designed. If that's an investment you think puts it a disadvantage, Vim probably isn't for you.
